Error with os-tenant-networks

I think a recent change in v0.2.0 caused a weird error for me.

I was able to do `terraform plan` and `terraform apply` to bring up a VM on openstack successfully. I then tried to do `terraform destroy`, and got this error:

```
openstack_compute_instance_v2.instance: Refreshing state... (ID: 732f11df-e122-4918-9554-42eff4de533e)
openstack_compute_floatingip_v2.ip: Refreshing state... (ID: 833)
Error refreshing state: 1 error(s) occurred:

* module.fe_master.module.compute.openstack_compute_instance_v2.instance: 1 error(s) occurred:

* module.fe_master.module.compute.openstack_compute_instance_v2.instance: openstack_compute_instance_v2.instance: Error trying to get network information from the Nova API: An error occured while querying the Nova API for network information: Expected HTTP response code [200 204] when accessing [GET https://openstack.url/v2.1/ec64afd6f5e547bb8b1a503cc08e2c7f/os-tenant-networks], but got 403 instead
{"forbidden": {"message": "Policy doesn't allow os_compute_api:os-tenant-networks to be performed.", "code": 403}}
```

Then when I do `terraform plan` again, I see the same error.

I checked with our cloud infrastructure team and they said that they do not allow access on `os_compute_api:os-tenant-networks` because they don't want me to delete tenant networks.

Is this a bug? I just want to delete the compute instance and floating IP, but not the network.

I think this might be due to https://github.com/terraform-providers/terraform-provider-openstack/commit/52adfae4febe4dba1cde5652fd382d9b2e5a69e5, but I'm not sure.

### Terraform Version
`v0.10.1`

### Affected Resource(s)
Please list the resources as a list, for example:
- `openstack_compute_instance_v1.instance`

When trying to do `terraform apply` again, I'm now getting this following error followed by "TERRAFORM CRASH"

```
2017-08-30T18:54:14.476-0400 [ERROR] plugin.terraform-provider-openstack_v0.2.1_x4: error parsing json from plugin: error="json: cannot unmarshal string into Go value of type map[string]interface {}"
```

I'm not sure what would cause it. Am heading out of the office for now, will try again tomorrow. If you have any suggestion, please let me know.

I get this error trying to do that

```
* openstack_networking_floatingip_v2.floatip_1: Error creating OpenStack network client: No suitable endpoint could be found in the service catalog.
2017/09/07 12:09:59 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:
```

--- a/openstack/compute_instance_v2_networking.go
+++ b/openstack/compute_instance_v2_networking.go
@@ -123,11 +123,15 @@ func getAllInstanceNetworks(d *schema.ResourceData, meta interface{}) ([]Instanc
 		}
 
 		v := InstanceNetwork{
-			UUID:          networkInfo["uuid"].(string),
-			Name:          networkInfo["name"].(string),
 			Port:          portID,
 			FixedIP:       network["fixed_ip_v4"].(string),
 			AccessNetwork: network["access_network"].(bool),
+		}
+		if networkInfo["uuid"] != nil {
+			v.UUID = networkInfo["uuid"].(string)
+		}
+		if networkInfo["name"] != nil {
+			v.Name = networkInfo["name"].(string)
 		}
 
 		instanceNetworks = append(instanceNetworks, v)
@@ -186,6 +190,39 @@ func getInstanceNetworkInfoNovaNet(
 	if queryType == "port" {
 		return nil, fmt.Errorf(
 			"Unable to query a port (%s) using the Nova API", queryTerm)
+	}
+
+	// test to see if the tenantnetworks api is available
+	log.Printf("[DEBUG] testing for os-tenant-networks")
+	tenantNetworksAvailable := true
+	if _, err := tenantnetworks.List(client).AllPages(); err != nil {
+		if _, ok := err.(gophercloud.ErrDefault404); ok {
+			log.Printf("[DEBUG] os-tenant-networks disabled")
+			tenantNetworksAvailable = false
+		}
+		log.Printf("[DEBUG] Err(type: %T) looks like: %+v", err, err)
+		switch errType := err.(type) {
+		case gophercloud.ErrUnexpectedResponseCode:
+			if errType.Actual == 403 {
+				log.Printf("[DEBUG] os-tenant-networks disabled.")
+				tenantNetworksAvailable = false
+			} else {
+				log.Printf("[DEBUG] unexpected os-tenant-networks error: %s", err)
+				tenantNetworksAvailable = false
+			}
+		case gophercloud.ErrDefault403:
+			log.Printf("[DEBUG] os-tenant-networks disabled.")
+			tenantNetworksAvailable = false
+		default:
+			log.Printf("[DEBUG] unexpected os-tenant-networks error: %s", err)
+			tenantNetworksAvailable = false
+		}
+	}
+
+	if !tenantNetworksAvailable {
+		// we can't query the APIs for more information, but in some cases
+		// the information provided is enough
+		return map[string]interface{}{queryType: queryTerm}, nil
 	}
 
 	allPages, err := tenantnetworks.List(client).AllPages()

--- a/openstack/compute_instance_v2_networking.go
+++ b/openstack/compute_instance_v2_networking.go
@@ -195,40 +195,27 @@ func getInstanceNetworkInfoNovaNet(
 	// test to see if the tenantnetworks api is available
 	log.Printf("[DEBUG] testing for os-tenant-networks")
 	tenantNetworksAvailable := true
-	if _, err := tenantnetworks.List(client).AllPages(); err != nil {
-		if _, ok := err.(gophercloud.ErrDefault404); ok {
-			log.Printf("[DEBUG] os-tenant-networks disabled")
+
+	allPages, err := tenantnetworks.List(client).AllPages()
+	if err != nil {
+		switch err.(type) {
+		case gophercloud.ErrDefault404:
 			tenantNetworksAvailable = false
-		}
-		log.Printf("[DEBUG] Err(type: %T) looks like: %+v", err, err)
-		switch errType := err.(type) {
-		case gophercloud.ErrUnexpectedResponseCode:
-			if errType.Actual == 403 {
-				log.Printf("[DEBUG] os-tenant-networks disabled.")
-				tenantNetworksAvailable = false
-			} else {
-				log.Printf("[DEBUG] unexpected os-tenant-networks error: %s", err)
-				tenantNetworksAvailable = false
-			}
 		case gophercloud.ErrDefault403:
-			log.Printf("[DEBUG] os-tenant-networks disabled.")
+			tenantNetworksAvailable = false
+		case gophercloud.ErrUnexpectedResponseCode:
 			tenantNetworksAvailable = false
 		default:
-			log.Printf("[DEBUG] unexpected os-tenant-networks error: %s", err)
-			tenantNetworksAvailable = false
+			return nil, fmt.Errorf(
+				"An error occurred while querying the Nova API for network information: %s", err)
 		}
 	}
 
 	if !tenantNetworksAvailable {
 		// we can't query the APIs for more information, but in some cases
 		// the information provided is enough
+		log.Printf("[DEBUG] os-tenant-networks disabled.")
 		return map[string]interface{}{queryType: queryTerm}, nil
-	}
-
-	allPages, err := tenantnetworks.List(client).AllPages()
-	if err != nil {
-		return nil, fmt.Errorf(
-			"An error occurred while querying the Nova API for network information: %s", err)
 	}
 
 	networkList, err := tenantnetworks.ExtractNetworks(allPages)

--- a/openstack/compute_instance_v2_networking.go
+++ b/openstack/compute_instance_v2_networking.go
@@ -458,7 +458,11 @@ func flattenInstanceNetworks(
 				if err != nil {
 					log.Printf("[WARN] Error getting default network uuid: %s", err)
 				} else {
-					v["uuid"] = networkInfo["uuid"].(string)
+					if v["uuid"] != nil {
+						v["uuid"] = networkInfo["uuid"].(string)
+					} else {
+						log.Printf("[WARN] Could not get default network uuid")
+					}
 				}
 
 				networks = append(networks, v)

